src/software/ai/passing/pass_generator.cpp

```diff
@@ -0,0 +1,174 @@
+        // velocity.
+        Point sampling_center =
+            robot_position + (robot.velocity() * sampling_center_vel_multiplier);
+        double std_dev = min_sampling_std_dev +
+                         (robot.velocity().length() * sampling_std_dev_vel_multiplier);
+        std::normal_distribution x_normal_distribution{sampling_center.x(), std_dev};
+        std::normal_distribution y_normal_distribution{sampling_center.y(), std_dev};
+
+        for (unsigned int i = 0; i < passing_config_.pass_gen_num_samples_per_robot();
+             i++)
+        {
+            auto point = Point(x_normal_distribution(random_num_gen_),
+                               y_normal_distribution(random_num_gen_));
+            receiving_positions_map[robot.id()].push_back(point);
+        }
+    }
+
+    return receiving_positions_map;
+}
+
+PassWithRating PassGenerator::optimizeReceivingPositions(
+    const World& world,
+    const std::map<RobotId, std::vector<Point>>& receiving_positions_map)
+{
+    // The objective function we minimize in gradient descent to improve each pass
+    // that we're optimizing
+    const auto objective_function =
+        [this, &world](const std::array<double, NUM_PARAMS_TO_OPTIMIZE>& pass_array) {
+            // get a pass with the new appropriate speed using the new destination
+            return ratePass(world,
+                            Pass::fromDestReceiveSpeed(
+                                world.ball().position(),
+                                Point(pass_array[0], pass_array[1]), passing_config_),
+                            passing_config_);
+        };
+
+    PassWithRating best_pass{Pass(Point(), Point(), 1.0), 0.0};
+    for (const auto& [robot_id, receiving_positions] : receiving_positions_map)
+    {
+        PassWithRating best_pass_for_robot{Pass(Point(), Point(), 1.0), 0.0};
+        for (const Point& receiving_position : receiving_positions)
+        {
+            auto optimized_receiving_pos_array = optimizer_.maximize(
+                objective_function, {receiving_position.x(), receiving_position.y()},
+                passing_config_.number_of_gradient_descent_steps_per_iter());
+
+            // get a pass with the new appropriate speed using the optimized destination
+            auto optimized_pass = Pass::fromDestReceiveSpeed(
+                world.ball().position(),
+                Point(optimized_receiving_pos_array[0], optimized_receiving_pos_array[1]),
+                passing_config_);
+            auto score = ratePass(world, optimized_pass, passing_config_);
+
+            if (score > best_pass_for_robot.rating)
+            {
+                best_pass_for_robot = PassWithRating{optimized_pass, score};
+            }
+        }
+
+        // Update the best receiving position for the robot used in the next iteration
+        // if the rating is above a certain threshold.
+        if (best_pass_for_robot.rating > 0.1)
```

Comment:
should this be a constant or a dynamic param?

Reply:
Honestly it's a pretty arbitrary value that I picked for basic filtering of passes from being re-evaluated in the next pass generator iteration. Not sure if making it a variable will make it much more clearer than what the comment explains